workflow/scripts/add_district_heating_subnodes.py

```diff
@@ -0,0 +1,393 @@
+# -*- coding: utf-8 -*-
+import logging
+
+logger = logging.getLogger(__name__)
+from random import randint
+from time import sleep
+
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+import pypsa
+import xarray as xr
+from geopy.geocoders import Nominatim
+
+
+# Function to encode city names in UTF-8
+def encode_utf8(city_name):
+    return city_name.encode("utf-8")
+
+
+def prepare_subnodes(subnodes, cities, regions_onshore, lau, heat_techs, head=40):
+    # TODO: Embed I&O in snakemake rule, add potentials, match CHP capacities
+    # If head is boolean set it to 40 for default behavior
+    if isinstance(head, bool):
+        head = 40
+
+    subnodes["Stadt"] = subnodes["Stadt"].str.split("_").str[0]
+
+    # Drop duplicates if Gelsenkirchen, Kiel, or Flensburg is included and keep the one with higher Wärmeeinspeisung in GWh/a
+    subnodes = subnodes.drop_duplicates(subset="Stadt", keep="first")
+
+    # Keep only n largest district heating networks according to head parameter
+    subnodes = subnodes.sort_values(
+        by="Wärmeeinspeisung in GWh/a", ascending=False
+    ).head(head)
+
+    subnodes["yearly_heat_demand_MWh"] = subnodes["Wärmeeinspeisung in GWh/a"] * 1e3
+
+    logger.info(
+        f"The selected district heating networks have an overall yearly heat demand of {subnodes['yearly_heat_demand_MWh'].sum()} MWh/a. "
+    )
+
+    subnodes["geometry"] = subnodes["Stadt"].apply(
+        lambda s: cities.loc[cities["Stadt"] == s, "geometry"].values[0]
+    )
+
+    subnodes = subnodes.dropna(subset=["geometry"])
+    # Convert the DataFrame to a GeoDataFrame
+    subnodes = gpd.GeoDataFrame(subnodes, crs="EPSG:4326")
+
+    # Assign cluster to subnodes according to onshore regions
+    subnodes["cluster"] = subnodes.apply(
+        lambda x: regions_onshore.geometry.contains(x.geometry).idxmax(), axis=1
+    )
+    subnodes["lau"] = subnodes.apply(
+        lambda x: lau.loc[lau.geometry.contains(x.geometry).idxmax(), "LAU_ID"], axis=1
+    )
+    subnodes["lau_shape"] = subnodes.apply(
+        lambda x: lau.loc[lau.geometry.contains(x.geometry).idxmax(), "geometry"].wkt,
+        axis=1,
+    )
+    subnodes["nuts3"] = subnodes.apply(
+        lambda x: heat_techs.geometry.contains(x.geometry).idxmax(),
+        axis=1,
+    )
+    subnodes["nuts3_shape"] = subnodes.apply(
+        lambda x: heat_techs.loc[
+            heat_techs.geometry.contains(x.geometry).idxmax(), "geometry"
+        ].wkt,
+        axis=1,
+    )
+
+    return subnodes
+
+
+def add_subnodes(n, subnodes):
```

Comment:
Do you mean nested functions? I would like to avoid them. Alternatively, I could split it up into `add_subnode_buses`, `add_subnode_loads`, `add_subnode_stores`, `add_subnode_links`, and `add_subnode_generators`. Might be a little bit overkill though?